### PR TITLE
feat: add Avoid unnecessary setState

### LINKED ISSUE #332: [New rule] avoid-unnecessary-setstate
**Please describe what the rule should do:**

As @GroovinChip pointed out in one of a twitter thread, calling `setState` from a widget `initState` or `build` methods will lead to unnecessary rerender of the widget. It looks like a potential problem especially for new developers, so the idea of the rule is to warn about `setState` invocations in `initState` or `build`.

We have not quite come to a conclusion, should we also consider calling sync methods from a widget `initState` with `setState` call in it bad or acceptable.

@GroovinChip thanks again for providing examples and sharing the idea.

Additional resources: 
* https://stackoverflow.com/questions/53363774/importance-of-calling-setstate-inside-initstate/53373017#53373017

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**
Bad:
```dart
class MyWidget extends StatefulWidget {
  @override
  _MyWidgetState createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  String myString = '';
  
  @override
  void initState() {
    super.initState();
   
    setState(() {
      myString = "Hello";
    }); // Bad
    
    if (condition) {
      setState(() {}); // Bad
    }
  }
  
  @override
  Widget build(BuildContext context) {
    setState(() {
      myString = "Hello";
    }); // Bad
    
    if (condition) {
      setState(() {}); // Bad
    }

    return ElevatedButton(
      child: Text('PRESS'),
    );
  }
}
```

```dart
class MyWidget extends StatefulWidget {
 @override
  _MyWidgetState createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  String myString = '';
  
  @override
  void initState() {
    super.initState();
    myMethod(); // Good or Bad ???
    myAsyncMethod(); // Good
  }
  
  void myMethod() {
    setState((){
      myString = "Hello";
    }); // Good
  }
 
  Future<void> myAsyncMethod() async {
    final data = await service.fetchData();
      
    setState(() {
      myString = data;
    });
  }
  
  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      onPressed: () => myMethod(), // Good
      child: Text('PRESS'),
    );
  }
}
```

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[X] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### What changes did you make? (Give an overview)
Closes #332

### CODE REVIEW

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      ..._visitor.setStateInvocations
+          .map((invocation) => createIssue(
+                rule: this,
+                location: nodeLocation(
+                  node: invocation,
+                  source: source,
+                ),
+                message: _warningMessage,
+              ))
+          .toList(growable: false),
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart:49]: @incendial please remove unnecessary toList.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      ..._visitor.classMethodsInvocations
+          .map((invocation) => createIssue(
+                rule: this,
+                location: nodeLocation(
+                  node: invocation,
+                  source: source,
+                ),
+                message: _methodWarningMessage,
+              ))
+          .toList(growable: false)
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart:59]: @incendial please remove unnecessary toList.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  static const _warningMessage = 'Avoid calling unnecessary setState.';
+  static const _methodWarningMessage = 'Avoid calling a method with setState.';
+
+  AvoidUnnecessarySetStateRule([Map<String, Object> config = const {}])
+      : super(
+          id: ruleId,
+          documentation: const RuleDocumentation(
+            name: 'Avoid returning widgets',
+            brief: 'Warns ',
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart:27]: @incendial please add more detailed message

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart (hunk 2 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart]: Done.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart (hunk 3 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart]: Done.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart (hunk 1 above)
Comment (incendial) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_unnecessary_setstate/avoid_unnecessary_setstate.dart]: Done.

**Review by dkrutskikh — APPROVED** [review]

**Pull request merged**